### PR TITLE
Add option to save real tensors in TORCH_COMPILE_DEBUG repro

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138110

This pr adds a utility to try to try to construct the corresponding real tensor values of fake tensors by seeing if their meta storage is contained in the meta converter. 

Then, we are able to save real tensor values for fx_graph_runnable if `TORCH_COMPILE_DEBUG_SAVE_REAL=1` is set.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

Differential Revision: [D64502744](https://our.internmc.facebook.com/intern/diff/D64502744)

### CODE REVIEW

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64502744).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64502744).

File: torch/_dynamo/debug_utils.py [hunk 1]
```diff
@@ -511,7 +511,7 @@ def symint(self, *args, **kwargs):
 # TODO: Support bundling the entire repro into a zip file for ease of
 # transferring around
 class InputReader:
-    def __init__(self, save_dir=None, *, pbar=None):
+    def __init__(self, save_dir=None, *, pbar=None, stable_hash=False):
```
Comment (ezyang) [inline torch/_dynamo/debug_utils.py:514]: Argument does nothing?

File: torch/_inductor/debug.py [hunk 2]
```diff
@@ -473,7 +473,26 @@ def fx_graph(
...
+            with torch._inductor.config.patch(
+                {"trace.enabled": False, "trace.save_real_tensors": False}
+            ):
+                save_graph_repro(
+                    fd,
+                    gm,
+                    inputs,
+                    "inductor",
+                    save_dir=save_dir,
+                    stable_hash=stable_hash,
```
Comment (ezyang) [inline torch/_inductor/debug.py:494]: Not sure I understand this? You only ever pass True here when this kwarg is actually respected?

File: torch/_subclasses/fake_utils.py [hunk 3]
```diff
@@ -75,6 +76,74 @@ def is_sdpa_error(func, idx, e):
     return False
 
 
+def try_convert_fake_to_real(
+    ten_list: List[Union[FakeTensor, Any]]
+) -> List[Union[FakeTensor, torch.Tensor, Any]]:
+    """
+    Attempt to convert fake tensors to a corresponding real tensor with the correct underlying storage by looking up
```
Comment (ezyang) [inline torch/_subclasses/fake_utils.py:83]: Can you explain why you have to do something this complicated? A trivial alternate implementation would be to have fake tensor maintain weakref to the real tensor they were converted from and read it off from that. Why do we have to do something so fancy?

**Review by ezyang — APPROVED** [review]

File: torch/_subclasses/fake_utils.py (hunk 3 above)
Comment (eellison) [reply torch/_subclasses/fake_utils.py:83]: Aot autograd will do weird things with views and such. this is more reliable.

**Pull request closed without merging**